onnxruntime/core/graph/graph_utils.cc

```diff
@@ -77,25 +77,128 @@ bool IsSingleInSingleOutNode(const Node& node) {
   return node.GetInputEdgesCount() == 1 && node.GetOutputEdgesCount() == 1;
 }
 
-const ONNX_NAMESPACE::AttributeProto* GetNodeAttribute(
-    const Node& node, const std::string& attr_name) {
+const ONNX_NAMESPACE::AttributeProto* GetNodeAttribute(const Node& node, const std::string& attr_name) {
   const auto& attrs = node.GetAttributes();
   const auto iter = attrs.find(attr_name);
   return iter == attrs.end() ? nullptr : &iter->second;
 }
 
+// check if new_output_name can be used to replace removed_output_name
+// if there is an existing NodeArg in a subgraph that implicitly consumes removed_output_name, it is not safe.
+static bool CanUpdateImplicitInputNameInSubgraph(Node& node,
+                                                 const std::string& removed_output_name,
+                                                 const std::string& new_output_name) {
+  for (auto& attr_subgraph_pair : node.GetAttributeNameToMutableSubgraphMap()) {
+    Graph& subgraph = *attr_subgraph_pair.second;
+    // if we have an existing NodeArg in the subgraph with the new_output_name that would override an implicit input
+    // with the same name
+    if (subgraph.GetNodeArg(new_output_name) != nullptr) {
+      return false;
+    }
+
+    for (auto& subgraph_node : attr_subgraph_pair.second->Nodes()) {
+      // recurse if this node also consumes removed_output_name as an implicit input (i.e. there are multiple levels of nested
+      // subgraphs, and at least one level lower uses removed_output_name as an implicit input
+      const auto& subgraph_node_implicit_inputs = subgraph_node.ImplicitInputDefs();
```

Comment:
If you are multiple levels of nesting in, will the ImplicitInputDefs() give you implicit inputs from any level above?

Reply:
Yes. We connect the input nodes with the control flow node at all levels required. If we didn't do that the topological sort would potentially break. See https://github.com/Microsoft/onnxruntime/blob/e9b61d90142483670ecc8fe4f27fb057fcfdbb03/onnxruntime/core/graph/graph.cc#L897